applications/serial_lte_modem/src/slm_cmux.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include "slm_cmux.h"
+#include "slm_at_host.h"
+#include "slm_uart_handler.h"
+#include <zephyr/logging/log.h>
+#include <zephyr/modem/backend/uart.h>
+#include <zephyr/modem/cmux.h>
+#include <zephyr/modem/pipe.h>
+#include <assert.h>
+
+LOG_MODULE_REGISTER(slm_cmux, CONFIG_SLM_LOG_LEVEL);
+
+/* This makes use of some of the Zephyr modem subsystem that has a CMUX module implementation.
+ */
+
+static struct {
+	/* UART backend */
+	struct modem_pipe *uart_pipe;
+	struct modem_backend_uart uart_backend;
+	uint8_t uart_backend_receive_buf[512];
+	uint8_t uart_backend_transmit_buf[512];
+
+	/* CMUX */
+	struct modem_cmux instance;
+	uint8_t cmux_receive_buf[128];
+	uint8_t cmux_transmit_buf[256];
+
+	/* CMUX channels (Data Link Connection Identifier); index = address - 1 */
+	struct cmux_dlci {
+		struct modem_cmux_dlci instance;
+		struct modem_pipe *pipe;
+		uint8_t *recv_buf;
+		uint16_t recv_buf_size;
+		uint8_t address;
+	} dlcis[2];
+} cmux;
+
+static void dlci_pipe_event_handler(struct modem_pipe *pipe,
+				    enum modem_pipe_event event, void *user_data)
+{
+	const struct cmux_dlci *const dlci = user_data;
+
+	switch (event) {
+	case MODEM_PIPE_EVENT_OPENED:
+	case MODEM_PIPE_EVENT_CLOSED:
+		LOG_INF("DLCI %u %s.", dlci->address,
+			(event == MODEM_PIPE_EVENT_OPENED) ? "opened" : "closed");
+		break;
+	case MODEM_PIPE_EVENT_RECEIVE_READY:
+		static uint8_t recv_buf[256];
+		const int recv_len = modem_pipe_receive(pipe, recv_buf, sizeof(recv_buf));
+
+		if (recv_len > 0) {
+			LOG_DBG("Received %u bytes on DLCI %u.", recv_len, dlci->address);
+			modem_pipe_transmit(pipe, recv_buf, recv_len);
+		} else {
+			LOG_ERR("Failed to receive data from DLCI %u (%d).",
+				dlci->address, recv_len);
+		}
+		break;
+	}
+}
+
+static void cmux_event_handler(struct modem_cmux *, enum modem_cmux_event event, void *)
+{
+	assert(event == MODEM_CMUX_EVENT_CONNECTED || event == MODEM_CMUX_EVENT_DISCONNECTED);
+	LOG_INF("CMUX %sconnected.", (event == MODEM_CMUX_EVENT_CONNECTED) ? "" : "dis");
+}
+
+static void init_dlci(size_t address, uint16_t recv_buf_size,
+		      uint8_t recv_buf[static recv_buf_size])
+{
+	assert(ARRAY_SIZE(cmux.dlcis) >= address);
+
+	struct cmux_dlci *const dlci = &cmux.dlcis[address - 1];
+	const struct modem_cmux_dlci_config dlci_config = {
+		.dlci_address = address,
+		.receive_buf = recv_buf,
+		.receive_buf_size = recv_buf_size
+	};
+
+	dlci->pipe = modem_cmux_dlci_init(&cmux.instance, &dlci->instance, &dlci_config);
+	dlci->recv_buf = recv_buf;
+	dlci->recv_buf_size = recv_buf_size;
+	dlci->address = address;
+
+	modem_pipe_attach(dlci->pipe, dlci_pipe_event_handler, dlci);
+}
+
+static int cmux_start(void)
+{
+	int ret;
+
+	/* Disable SLM UART so that CMUX can take over it. */
+	ret = slm_uart_rx_disable();
```

Comment:
Note: This does prevent TX, so something that is already running could theoretically be sending. Given that you have disabled most of the clients with PPP, that is unlikely to be a problem. But something to stay aware. 

Mechanism to disable TX would be reasonably simple, but unnecessary unless this is observed to cause problems.

Reply:
The (end) goal is to have both RX and TX go through CMUX. As of this PR the TX part is indeed not handled, as this is just echoing back what is received.
Though maybe `slm_uart_rx_disable()` should be renamed? Do you have a suggestion?

Reply:
I think that the name is in line what it does, so no need to remove at this point.